Fixed DbSession for Microsoft SQL Server, enhanced tests

SilverFire opened the pull request:
Fixes #13436

| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13436

![screen shot 2017-10-30 at 15 20 24](https://user-images.githubusercontent.com/4499203/32172988-2d4a74aa-bd86-11e7-8f03-b3daee292e2c.png)
![screen shot 2017-10-30 at 15 20 56](https://user-images.githubusercontent.com/4499203/32172989-2d69db4c-bd86-11e7-9932-0addca2d0cc7.png)
![screen shot 2017-10-30 at 15 21 09](https://user-images.githubusercontent.com/4499203/32172990-2d915942-bd86-11e7-9628-ec3056814b0e.png)
![screen shot 2017-10-30 at 15 22 29](https://user-images.githubusercontent.com/4499203/32172992-2dd960c0-bd86-11e7-8ed7-f13b4e9149b8.png)

SilverFire committed:
Fixed DbSession for Microsoft SQL Server, enhanced tests

Fixes #13436

cebe commented on tests/framework/web/session/DbSessionTest.php:
```diff
@@ -16,42 +16,74 @@
  * @group db
+ * @group sqlite
  */
 class DbSessionTest extends TestCase
```
better make this class abstract and create an sqlite implementation of it.

cebe commented on framework/web/migrations/m160313_153426_session_init.php:
```diff
@@ -20,16 +20,23 @@ class m160313_153426_session_init extends Migration
+                $tableOptions = 'CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE=InnoDB';
+                break;
+            case 'sqlsrv':
+                $dataType = $this->text();
```
we had issues with non-utf8 data stored in session, which caused text to fail so we used binary for that. Is this covered by a test?

SilverFire replied:
I remembered it, but was not sure about characters that cause that behavior, so typed something random: https://github.com/yiisoft/yii2/pull/15049/files#diff-722a201958cf84ca29390026ac2a10eaR138

Hope it covers. If you remember something more exact – share a link please

cebe replied:
The issue is about raw binary data that does not follow utf8 encoding, e.g. https://github.com/yiisoft/yii2/issues/933

SilverFire committed:
Moved sqlite test out of DbSessionTest. Renamed DbSessionTest to AbstractDbSessionTest

Fixed tests to work corrctly even in case previous test did not clean up migrations table

samdark commented:
Hmmm... fails.

SilverFire committed:
Fixed DbManagerTestCase to revert all migrations

Before this commit, DbManagerTestCase reverted only the latest migration, so test does not restore DB
after execution and may affect other tests

SilverFire commented:
Fixed `rbac/DbManagerTestCase` to revert all migrations, instead of the latest only

cebe reviewed (changes requested):
Need a test to verify raw binary data does not cause issues. See e.g. #933, which is about raw binary data created by encryption that does not follow utf8 encoding.

SilverFire committed:
Added DbSession::typecastFields() to handle null-byte in `data` correctly

SilverFire commented:
Yeah, it did not work. I've tried to fix it, but was able to fix only work with control chars (null-byte, etc).
Inserting binary data leads to the error that occurs either with `nvarchar(max)` or `varbinnary(max)`:

```
An error occurred translating string for input param 1 to UCS-2: Error code 0x7
```

I've adjusted test to make sure other DBMSs handle binary session data correctly, but I'm not a Microsoft SQL Server expert and I doubt I can fix any more.

Symfony [PdoSessionHandler test](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Tests/Session/Storage/Handler/PdoSessionHandlerTest.php) does not check possibility of inserting binary data and I think I know why ;)

cebe commented:
if we can't fix it its okay to merge it, but maybe @sergeymakinen has an idea how to make it work?

samdark commented:
He's extremely busy so waiting may hold release significantly.

samdark committed:
Merge branch 'master' into 13436-mssql-dbsession

samdark committed:
Removed unneeded lines [skip ci]
